File: web/movei-web/services/scanner-service.ts
import { supabase, isSupabaseConfigured } from '../lib/supabase';
import { ScanResponse } from '../types';
import { MOCK_TICKETS } from '../lib/mock-data';

export async function validateAndAdmitTicket(
  ticketCode: string,
  scannerId: string = 'u3333333-3333-3333-3333-333333333333',
  cinemaId: string = 'c1111111-1111-1111-1111-111111111111',
  deviceId?: string
): Promise<ScanResponse> {
  if (isSupabaseConfigured && supabase) {
    const { data, error } = await supabase.rpc('validate_and_admit_ticket', {
      p_ticket_code: ticketCode.trim(),
      p_scanner_id: scannerId,
      p_cinema_id: cinemaId,
      p_device_id: deviceId || 'web-browser-scanner'
    });
    if (error) {
      console.error('Scan RPC error:', error);
      return {
        valid: false,
        reason: 'invalid',
        message: error.message || 'Database error during scan'
      };
    }
    return data as ScanResponse;
  }

  // Local mock admission logic with atomic checks
  const ticket = MOCK_TICKETS.find(
    t => t.ticket_code.toUpperCase() === ticketCode.trim().toUpperCase() ||
         t.barcode_value.toUpperCase() === ticketCode.trim().toUpperCase()
  );

  if (!ticket) {
    return {
      valid: false,
      reason: 'invalid',
      message: 'Ticket not found in system'
    };
  }

  if (ticket.show?.cinema_id && ticket.show.cinema_id !== cinemaId) {
    return {
      valid: false,
      reason: 'wrong_cinema',
      message: `Ticket is for ${ticket.show?.cinema?.name || 'another cinema'}`
    };
  }

  if (ticket.status === 'used') {
    return {
      valid: false,
      reason: 'already_used',
      message: `Already used / admitted at ${ticket.scanned_at ? new Date(ticket.scanned_at).toLocaleTimeString() : 'earlier today'}`
    };
  }

  if (ticket.status === 'cancelled' || ticket.status === 'expired') {
    return {
      valid: false,
      reason: ticket.status,
      message: `Ticket is marked as ${ticket.status}`
    };
  }

  // Mark admitted
  ticket.status = 'used';
  ticket.scanned_at = new Date().toISOString();
  ticket.scanned_by = scannerId;

  return {
    valid: true,
    reason: 'valid',
    ticket_code: ticket.ticket_code,
    movie_title: ticket.show?.movie?.title || 'Unknown Title',
    customer_name: 'Alex Mercer',
    cinema_name: ticket.show?.cinema?.name || 'Cinema Complex',
    screen_name: ticket.show?.screen?.name || 'Screen 1',
    seat: `${ticket.seat?.row_label}${ticket.seat?.seat_number}`,
    showtime: ticket.show?.start_time
  };
}

File: MOVEI/MOVEI/Services/ScannerService.swift
//
//  ScannerService.swift
//  MOVEI
//

import SwiftUI
import Combine

public struct AdmissionResult {
    public let isValid: Bool
    public let statusType: StatusType
    public let title: String
    public let message: String
    public let ticketCode: String
    public let movieTitle: String?
    public let customerName: String?
    public let cinemaName: String?
    public let screenName: String?
    public let seat: String?
    public let showtime: Date?
    public let scannedAt: Date?

    public enum StatusType {
        case valid
        case alreadyUsed
        case wrongCinema
        case cancelled
        case expired
        case invalid
    }
}

@MainActor
public final class ScannerService: ObservableObject {
    public static let shared = ScannerService()

    @Published public var assignedCinema: Cinema = Cinema(
        id: "cinemax-colombo",
        name: "Cinemax Colombo",
        address: "125 Galle Road, Colombo 03"
    )
    @Published public var recentScans: [TicketScan] = []
    @Published public var totalAdmissionsToday: Int = 24

    private init() {
        // Preload recent scan logs for scanner dashboard
        self.recentScans = [
            TicketScan(ticketCode: "MOV-82K9A-01", scannerID: "scanner-1", cinemaID: "cinemax-colombo", scannedAt: Date().addingTimeInterval(-1800), result: "valid", movieTitle: "Wicked", customerName: "Sadew", seat: "B4"),
            TicketScan(ticketCode: "MOV-82K9A-02", scannerID: "scanner-1", cinemaID: "cinemax-colombo", scannedAt: Date().addingTimeInterval(-1780), result: "valid", movieTitle: "Wicked", customerName: "Sadew", seat: "B5"),
            TicketScan(ticketCode: "MOV-19FB02-01", scannerID: "scanner-1", cinemaID: "cinemax-colombo", scannedAt: Date().addingTimeInterval(-720), result: "already_used", movieTitle: "Oppenheimer", customerName: "Michael", seat: "D3")
        ]
    }

    public func validateTicket(barcode: String) -> AdmissionResult {
        let cleanCode = barcode.trimmingCharacters(in: .whitespacesAndNewlines)
        let ticketService = TicketService.shared

        guard let ticket = ticketService.tickets.first(where: { $0.ticketCode == cleanCode || $0.barcodeValue == cleanCode }) else {
            let scan = TicketScan(ticketCode: cleanCode, scannerID: "scanner-04", cinemaID: assignedCinema.id, result: "invalid")
            recentScans.insert(scan, at: 0)
            return AdmissionResult(
                isValid: false,
                statusType: .invalid,
                title: "INVALID TICKET",
                message: "No ticket record found matching code: \(cleanCode)",
                ticketCode: cleanCode,
                movieTitle: nil,
                customerName: nil,
                cinemaName: nil,
                screenName: nil,
                seat: nil,
                showtime: nil,
                scannedAt: nil
            )
        }

        // Venue check
        if !ticket.cinemaName.localizedCaseInsensitiveContains(assignedCinema.name) &&
           !assignedCinema.name.localizedCaseInsensitiveContains(ticket.cinemaName) {
            let scan = TicketScan(ticketID: ticket.id, ticketCode: cleanCode, scannerID: "scanner-04", cinemaID: assignedCinema.id, result: "wrong_cinema", movieTitle: ticket.movieTitle, customerName: "Customer", seat: ticket.seatLabel)
            recentScans.insert(scan, at: 0)
            return AdmissionResult(
                isValid: false,
                statusType: .wrongCinema,
                title: "WRONG VENUE",
                message: "Ticket is valid only at: \(ticket.cinemaName)",
                ticketCode: cleanCode,
                movieTitle: ticket.movieTitle,
                customerName: nil,
                cinemaName: ticket.cinemaName,
                screenName: ticket.screenName,
                seat: ticket.seatLabel,
                showtime: ticket.showtime,
                scannedAt: nil
            )
        }

        // Duplicate entry check
        if ticket.status == "used" {
            let scan = TicketScan(ticketID: ticket.id, ticketCode: cleanCode, scannerID: "scanner-04", cinemaID: assignedCinema.id, result: "already_used", movieTitle: ticket.movieTitle, customerName: "Customer", seat: ticket.seatLabel)
            recentScans.insert(scan, at: 0)
            return AdmissionResult(
                isValid: false,
                statusType: .alreadyUsed,
                title: "ALREADY USED",
                message: "Ticket was already admitted at \(ticket.scannedAt?.formatted(date: .omitted, time: .shortened) ?? "earlier").",
                ticketCode: cleanCode,
                movieTitle: ticket.movieTitle,
                customerName: nil,
                cinemaName: ticket.cinemaName,
                screenName: ticket.screenName,
                seat: ticket.seatLabel,
                showtime: ticket.showtime,
                scannedAt: ticket.scannedAt
            )
        }

        // Valid admission
        ticketService.markTicketUsed(ticketCode: cleanCode)
        totalAdmissionsToday += 1
        let scan = TicketScan(ticketID: ticket.id, ticketCode: cleanCode, scannerID: "scanner-04", cinemaID: assignedCinema.id, result: "valid", movieTitle: ticket.movieTitle, customerName: "Sadew", seat: ticket.seatLabel)
        recentScans.insert(scan, at: 0)
        UINotificationFeedbackGenerator().notificationOccurred(.success)

        return AdmissionResult(
            isValid: true,
            statusType: .valid,
            title: "VALID TICKET",
            message: "Admission authorized. Enjoy the movie!",
            ticketCode: cleanCode,
            movieTitle: ticket.movieTitle,
            customerName: "Sadew",
            cinemaName: ticket.cinemaName,
            screenName: ticket.screenName,
            seat: ticket.seatLabel,
            showtime: ticket.showtime,
            scannedAt: Date()
        )
    }
}
